ops/build_containers.py

```diff
@@ -0,0 +1,96 @@
+from os import path, curdir
+from subprocess import run
+from tempfile import gettempdir
+import urllib.request
+from json import loads
+
+from config import CONFIG
+from utils import log, checkoutRepo
+
+dockerApiUrl="https://hub.docker.com/v2/repositories"
+
+def doesContainerExistsInCloud(name: str, tag: str) -> bool:
+    try:
+        urllib.request.urlopen("{}/{}/tags/{}".format(dockerApiUrl, name, tag)).read()
+        return True
+    except urllib.error.HTTPError:
+        return False
+
+def getContainerInfo(name: str, tag: str) -> str:
+    if doesContainerExistsInCloud(name, tag):
+        url = "{}/{}/tags/{}".format(dockerApiUrl, name, tag)
+        return loads(urllib.request.urlopen(url).read().decode())
+        
+    raise urllib.error.URLError('Container {}:{} not found in registry'.format(name, tag))
+
+def compareRemoteContainers(containerOne: str, containerTwo: str) -> bool:
+    try:
+        info1 = getContainerInfo(containerOne.split(":")[0], containerOne.split(":")[1])
+        info2 = getContainerInfo(containerTwo.split(":")[0], containerTwo.split(":")[1])
+        return info1['images'][0]['digest'] == info2['images'][0]['digest']
+    except urllib.error.URLError as e:
+        log(e.reason)
+        return False
+
+def buildContainer(container: str, tag: str, dockerFile: str, path: str) -> None:
+    log("Building {}:{}".format(container, tag))
+    args = [
+        "docker",
+        "build",
+        "-t",
+        "{}:{}".format(container, tag),
+        "-f",
+        dockerFile,
+        path,
+    ]
+    if CONFIG["WARPED"]:
+        args.extend([
+            "--build-arg",
+            "warped=true",
+            "--build-arg",
+            "server={}".format(CONFIG["REPOS"]["SERVER"]["NAMEDTAG"])
+        ])
+    log(args)
+    run(args, check=True)
```

Comment:
Graceful failure is not always something we want. I need to think better but failures make  sure the tests don't pass and we can take a look at what is happening.

Not saying that I'm against your proposal, I just need to look at it more and see where it is necessary.

Reply:
Maybe graceful failure isn't the term here but we also need to consider the fact that devops people will be using these scripts as well and it would make their lives easier if we have descriptive messages telling them what might be going wrong. We can't only consider integration tests.